whisper/convert.py

```diff
@@ -0,0 +1,103 @@
+# Copyright © 2023 Apple Inc.
+
+import argparse
+import copy
+import json
+from dataclasses import asdict
+from pathlib import Path
+
+import mlx.core as mx
+import mlx.nn as nn
+import numpy as np
+from mlx.utils import tree_flatten, tree_map, tree_unflatten
+
+from whisper.load_models import load_torch_model, torch_to_mlx
+from whisper.torch_whisper import ModelDimensions
+from whisper.whisper import Whisper
+
+MODEL_DTYPES = {"float16", "float32"}
+
+def quantize(weights, config, dtype, args):
+    quantized_config = copy.deepcopy(config)
+
+    # Load the model:
+    model = Whisper(ModelDimensions(**config), dtype)
```

Comment:
Yeah, I was thinking of having an option to keep some non-linear layers in FP32. It seems unnecessary if we're already doing quantization.

Reply:
> Yeah, I was thinking of having an option to keep some non-linear layers in FP32

The dtype is used to construct the mask and the sinusoidal position encodings, (both of which are not parameters). Those don't even get saved / returned in the `model.paramters()`, so this really doesn't have any affect. All the other params are overwritten by the input weights when you do `model.udpate(...)` in the line below.

Reply:
I was thinking about the CNN layers since the weights are converted using `tree_map(lambda p: p.astype(dtype), params)` before being passed to models. But I get your point, and also believe it's basically the same thing.